doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,258 @@
+=========
+Pipelines
+=========
+
+RADOS Protocol Requirements
+===========================
+
+Requests
+--------
+
+OSD serves different kinds of requests. Some of them are targeting the local state of the connected
+OSD instance. For instance, the ``ceph config get`` command is directed straight to the local config
+subsystem, and is fulfilled right away. But others inquiry or mutate the shared state of the cluster.
+Following are two categories of these requests:
+
+- I/O requests from clients.
+- PG peering requests from peer OSDs.
+
+Because inherently a distributed storage system is distributed by its own nature, we can not assume
+that an OSD or the PGs hosted by it are always ready to serve a request targeting a shared state, not
+to mention the unpredictable hardware failures. But fortunately, even if an OSD is not able to serve
+a request right away, under most circumstances, it can postpone the request, prepare itself so that it
+can answer this request. It's quite common in daily life:
+
+  Great question, let me look into it and I will get back to you!
+
+Some questions involves multiple steps. To serve an I/O request from client, at
+least 7 stages are involved:
+
+.. ditaa::
+
+   +----+---------------------+
+   |cGRE| OSD wait for osdmap |
+   +----+----+----------------+
+             |
+             V
+   +----+-----------------+
+   |cGRE| OSD wait for PG |
+   +----+-----------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for osdmap |
+   +----+--------------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for active |
+   +----+--------------------+
+             |
+             V
+   +----+----------------------+
+   |cBLU| PG wait for recovery |
+   +----+----------------------+
+             |
+             V
+   +----+---------------------+
+   |cBLU| wait for object ctx |
+   +----+---------------------+
+             |
+             V
+   +----+------------------+
+   |cBLU| wait for process |
+   +----+------------------+
+
+             legend
+             +----------------+
+             |cGRE OSD stages |
+             +----------------+
+             |cBLU PG stage   |
+             +----------------+
+
+Two Requirements
+----------------
+
+An OSD client talks with each connected OSD over stateful connection. The client is allowed to
+perform asynchronous I/O with OSD in the sense that it can send another request without waiting
+for the completion of the previous one. But the contract of the RADOS protocol requires writes
+(and write ordered reads) performed on a single object to be performed in the order in which
+they were originally submitted, and only once. This behavior is implemented by both the client
+side and server side of librados.
+
+So far, we have two restrictions:
+
+#. sequential stages of a given request
```

Comment:
Sequential stages aren't required, but they are sufficient.  This choice could be softened at the expense of substantial complexity.

Reply:
yes. i remove this item. as it's not required but a good approach to structure the checks and the preparations for satisfy the preconditions.